meeteval/io/py.py

```diff
@@ -0,0 +1,254 @@
+            )
+
+    if len(types) == 1:
+        # After modification, it can happen that the SegLST representation contains multiple segments.
+        # We concatenate here to keep some sort of "old" behavior.
+        #
+        # Note: This inversion gives back the original when the SegLST representation is not modified,
+        # but it can give a different representation when it was modified.
+        # TODO: Would it be better to keep a list? That would make the above statement true, but it would
+        #  change the format, e.g., from str to List[str] and is not guaranteed to do the same at
+        #  all nesting levels
+        if keys[0] == 'words':
+            words = [s['words'] for s in t]
+            if types[0] == str:
+                return ' '.join(words)
+            else:
+                return types[0](words)
+        assert len(t) == 1, t
+        return types[0](t.segments[0][keys[0]])
+    groups = {k: _invert_python_structure(v, types[1:], keys[1:]) for k, v in t.groupby(keys[0]).items()}
+    if types[0] in (list, tuple):
+        groups = types[0](groups.values())
+    return groups
+
+
+@dataclasses.dataclass(frozen=True)
+class NestedStructure(BaseABC):
+    """
+    Wraps a Python structure where the structure levels represent keys.
+
+    Example structure for cpWER:
+        ```python
+        structure = {
+            'Alice': ['segment1', 'segment2'],   # Speaker 1: Alice
+            'Bob': ['segment1', 'segment2'],   # Speaker 2: Bob
+        }
+        NestedStructure(structure, ('speaker', 'segment'))
+        ```
+    """
+    structure: Any
+    level_keys: 'List[str, ...] | Tuple[str, ...]' = ('speaker', 'segment_index')
+    final_key: 'str' = 'words'
+    final_types: 'type | List[type] | Tuple[type]' = str
+
+    # Cache variables
+    _types = None
+    _used_keys = None
+    _seglst = None
+
+    def new(self, t: 'SegLST', **defaults) -> 'Self':
+        """
+        This is usually a classmethod, but here, it's an instance method
+        because we need `keys` and `types` for conversion.
+
+        >>> def convert_cycle(structure, keys, mod=None):
+        ...     s = NestedStructure(structure, keys)
+        ...     t = s.to_seglst()
+        ...     if mod:
+        ...         t = mod(t)
+        ...     return s.new(t).structure
+        >>> convert_cycle('a b c', keys=())
+        'a b c'
+        >>> convert_cycle(['a b c', 'd e f'], keys=('speaker',))
+        ['a b c', 'd e f']
+        >>> convert_cycle({'A': 'a b c', 'B': 'd e f'}, keys=('speaker',))
+        {'A': 'a b c', 'B': 'd e f'}
+        >>> s = NestedStructure({'B': 'd e f', 'A': 'a b c'}, level_keys=('speaker',))
+        >>> s2 = NestedStructure(['a b c', 'd e f'], level_keys=('speaker',))
+        >>> s.new(s2.to_seglst()).structure
+        {0: 'a b c', 1: 'd e f'}
+        >>> s2.new(s.to_seglst()).structure   # Keys are sorted when converting to list (TODO: is this expected behavior?)
```

Comment:
Since python 3.7 the order of dicts is part of the language (implemented since cpython 3.6). We shouldn't touch the order.
I see only reasons, there the sort is annoying, but no reason, where it improves the code.

Reply:
The reason is that if you change the keys, you want the order to change when you convert back to list. Example:

```Python
d = ['abc', 'def']
c = asseglistconvertible(d)
s = c.to_seglst()
s[0]['speaker'] = 1
s[1]['speaker'] = 0
c.new(s)
# Should this be
['abc', 'def']
# or 
['def', 'abc']
# ?
```

Reply:
Hmm. So with and without sort, the result can be unexpected:
 - `speaker` $\in$ `{'spk10,spk200,...'}`, if we sort them, then we have to use natsort
 - `speaker` $\in$ `{0, 1, ...}` the user expects a sort
 - `speaker` $\in$ `{'0', '1', ..., '10', ...}` should this be sorted? If yes, we have to use natsort

Do we need the support for unnamed speakers? It makes everything much more complicated.

Reply:
We need support for converting them to `SegLST` as not to break code that already uses the Python interface. The inversion is not important.

We could forbid non-int keys for conversion to lists with a good error message

Reply:
Yes, that is a good idea.

Reply:
Done